source/fab/tools/compiler.py

```diff
@@ -0,0 +1,301 @@
+##############################################################################
+# (c) Crown copyright Met Office. All rights reserved.
+# For further details please refer to the file COPYRIGHT
+# which you should have received as part of this distribution
+##############################################################################
+
+"""This file contains the base class for any compiler, and derived
+classes for gcc, gfortran, icc, ifort
+"""
+
+import os
+from pathlib import Path
+from typing import List, Optional, Union
+import zlib
+
+from fab.tools.categories import Categories
+from fab.tools.flags import Flags
+from fab.tools.tool import VendorTool
+
+
+class Compiler(VendorTool):
+    '''This is the base class for any compiler. It provides flags for
+
+    - compilation only (-c),
+    - naming the output file (-o),
+    - OpenMP
+
+    :param name: name of the compiler.
+    :param exec_name: name of the executable to start.
+    :param vendor: name of the compiler vendor.
+    :param category: the Category (C_COMPILER or FORTRAN_COMPILER).
+    :param compile_flag: the compilation flag to use when only requesting
+        compilation (not linking).
+    :param output_flag: the compilation flag to use to indicate the name
+        of the output file
+    :param omp_flag: the flag to use to enable OpenMP
+    '''
+
+    # pylint: disable=too-many-arguments
+    def __init__(self, name: str, exec_name: str, vendor: str,
+                 category: Categories,
+                 compile_flag: Optional[str] = None,
+                 output_flag: Optional[str] = None,
+                 omp_flag: Optional[str] = None):
+        super().__init__(name, exec_name, vendor, category)
+        self._version = None
+        self._compile_flag = compile_flag if compile_flag else "-c"
+        self._output_flag = output_flag if output_flag else "-o"
+        self._omp_flag = omp_flag
+        self.flags.extend(os.getenv("FFLAGS", "").split())
+
+    def get_hash(self) -> int:
+        ''':returns: a hash based on the compiler name and version.
+        '''
+        return (zlib.crc32(self.name.encode()) +
```

Comment:
I don't think we should hash() - afaik the python has values are randomly (i.e. 'salted' at startup with a random offset, to avoid a potential DOS attach by forcing hash values clusters), which would mean that the hash for anything would change from run to run.

Reply:
I've raised ticket #317 to investigate this.

Reply:
I've answered :)